### ISSUE TITLE
Would very much like to get javascript working

### ISSUE THREAD

**Issue by joequant** [hyperledger-archives/fabric#807]
We have some smart contracts that are written in javascript and would very much like to help with the coding of getting nodejs chaincode working.  What do we need to do?

**Comment by muralisrini** [comment]
@joequant : at a high level you'll have to do the following
1. convert the shim to javascript. This is code under obc-peer/openchain/chaincode/shim. It entails two main parts : (a) generating gRPC for javascript and (b) converting code to javascript
2. adding a switch for javascript  to the deploy code path (currently supports "golang")
1. is core and by far the  more complicated of the two tasks. It is better to do it first, separate from 2. And it is easier done by running  obc-peer in "peer-chaincodedev" mode as that removes need to run in container and makes it easy to minimize dependencies with 2.

Hope this helps to get started.

**Comment by binhn** [comment]
@mastersingh24 you're looking into this at one point

**Comment by joequant** [comment]
OK.  I'll take a look

The reason that I'm interested is that I have "smart contracts" that I'd like to have work in openblockchain

https://github.com/joequant/bitquant/tree/master/smart-contracts

**Comment by mastersingh24** [comment]
@joequant  - you still need some help with this?  I have a few folks that might be willing to help.  We've already done a bunch of work with GRPC in JavaScript.  Let me know if still interested in this and/or need help.  I personally am a big fan of JavaScript myself.

Another idea I toyed with was actually looking at using an embedded JavaScript engine for Go as well but I see a lot of dangers in that

**Comment by joequant** [comment]
Yes.  I'm interested.  What I'd like to see is if someone can sketch out
what code needs to be written, I can have a go at writing part of the code.

I looked at otto.  Right now, it seems that all you need to do is to
replicate the REST API to the docker apps?  Correct?

**Comment by mastersingh24** [comment]
@joequant  - okay - we'll take a stab at getting a basic shell of the Javascript SHIM done in the next couple of days and help point you in the right direction

**Comment by srderson** [comment]
@joequant @mastersingh24 any updates?

**Comment by smithbk** [comment]
See https://github.com/hyperledger/fabric/tree/master/sdk/node

**Comment by joequant** [comment]
Is that just a client api or can it be used as a shim for writing chaincode in javascript?

**Comment by smithbk** [comment]
@joequant Yes, it is just a client API and not for chaincode.  Should have looked more closely at request.

**Comment by joequant** [comment]
Also how does this relate to the ibm-blockchain-js code?  Which is the
recommended mechanism for nodejs programming?

**Comment by christo4ferris** [comment]
@joequant there's a project to develop a Java shim, to complement the Golang shim. There are now two examples of what is needed. I'm sure there are likely others interested in this feature. Suggest that you take this to the hyperledger-fabric mailing list to sus-out who else is interested, and see whether there is interest in developing/collaborating on a javascript shim.
